### PR TITLE
Add serviceWorker option

### PR DESCRIPTION
**Pull request by perrin4869**
This is a rather naive implementation of the API I proposed in #56. For now the event handlers

```javascript
this.myNotify.addEventListener('show', this, false);
this.myNotify.addEventListener('error', this, false);
this.myNotify.addEventListener('close', this, false);
this.myNotify.addEventListener('click', this, false);
```

don't work. Also, https://github.com/alexgibson/notify.js/compare/master...perrin4869:master#diff-bd33361ada7e8d2d2fefd6aa1b7a6061R167 should probably be improved.

Other than that, it works just fine. Please test this and give me your feedback :)

### CODE REVIEW

File: notify.js [hunk 1]
```diff
@@ -10,24 +10,32 @@
     if (typeof define === 'function' && define.amd) {
         // AMD environment
         define(function() {
-            return factory(global, global.document);
+            return factory(global, global.document, global.navigator);
```
Comment (alexgibson) [inline notify.js]: Why is this needed?

File: notify.js [hunk 2]
```diff
@@ -99,6 +108,11 @@
         try {
             new N('');
         } catch (e) {
+            if (SW) {
```
Comment (alexgibson) [inline notify.js]: This variable is only used once, so could you just do `if (w.navigator.serviceWorker)`?

Comment (alexgibson) [inline notify.js]: Also is this possibly a false positive? Supporting Service Worker does not necessarily mean supporting push notifications.

File: notify.js [hunk 3]
```diff
@@ -139,16 +153,30 @@
...
+            var self = this;
+
+            this.options.serviceWorker.ready.then(function(registration) {
+                registration.showNotification(title, options)
+                .then(function() {
+                    return registration.getNotifications();
+                })
+                .then(function(notifications) {
+                    self.myNotify = notifications[notifications.length - 1]; // TODO: Naive implementation, improve
+                    bindNotificationEvents(self);
```
Comment (alexgibson) [inline notify.js]: AFAIK, Push Notifications do not support the same events as web notifications. Please read the MDN docs for more detail.

**Comment by alexgibson** [comment]
Is this the right approach?

In #56 we proposed that Notify.js should be imported as a script in the Service Worker, which is then responsible for creating notifications. This seems to only create a notification once a SW is registered, which isn't quite what I'm looking for, unless I'm misunderstanding something?

**Comment by alexgibson** [comment]
Thanks for taking a first stab at this! It's great to see some progress, but I'd like to try and understand your approach a bit more re the question above :)

File: notify.js (hunk 3 above)
Comment (alexgibson) [inline notify.js]: This might be what you're looking for https://developer.mozilla.org/en-US/docs/Web/API/NotificationEvent

File: notify.js [hunk 4]
```diff
@@ -10,24 +10,32 @@
...
     'use strict';
 
     var N = w.Notification;
+    var SW = n.serviceWorker;
 
     function isFunction(item) {
         return typeof item === 'function';
     }
 
+    function bindNotificationEvents(notify) {
```
Comment (alexgibson) [inline notify.js]: I'd probably favor adding this to the `prototype` if it's going to be a function?

File: notify.js (hunk 3 above)
Comment (perrin4869) [inline notify.js]: The thing with MDN is that there are some misleading things there. For example:

https://developer.mozilla.org/en-US/docs/Web/API/NotificationEvent/notification

Says that `NotificationEvent.notification` returns a `Notification` object:

https://developer.mozilla.org/en-US/docs/Web/API/Notification

However, the events supported in the `Notification` event do not seem to fire.
Another misleading thing was this:
https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerRegistration/showNotification

MDN says that the method returns a `Promise` that resolves to a `NotificationEvent`, however, that is not true, it resolves to `undefined`, and if you look at the actual specification here:

https://notifications.spec.whatwg.org/#service-worker-api

You see that the `Promise` resolves to `void`
You basically need to test these things one by one, can't rely on the documentation

File: notify.js (hunk 2 above)
Comment (perrin4869) [inline notify.js]: Well, this is the best I could come up with if we want a synchronous function... I agree with you, it's not perfect
Can totally just use `w.navigator.serviceWorker`, I just thought I'd try to follow your previous conventions

File: notify.js (hunk 1 above)
Comment (perrin4869) [inline notify.js]: trying to follow the conventions like I understood them, will remove it

File: notify.js (hunk 4 above)
Comment (perrin4869) [inline notify.js]: You're right there!

File: notify.js (hunk 3 above)
Comment (alexgibson) [inline notify.js]: You might be getting a bit mixed up maybe, it seems like the docs say this event should be on the SW object itself. The first paragraph says:

"The parameter passed into the onnotificationclick handler, the NotificationEvent interface represents a notification click event that is dispatched on the ServiceWorkerGlobalScope of a ServiceWorker."

Comment (perrin4869) [inline notify.js]: `onnotificationclick` is just an event fired when you click on the notification, similar to `Notification.onclick`. Indeed, it is attached to the serviceWorker object. It is completely unrelated to `show`, `close`, `error`, however. And also, somewhat different to `click`. I think on the ServiceWorker case, the first 3 should be dropped, and we can try to make `click` event somewhat transparent to the user, although I got to think if that is even possible.

Comment (alexgibson) [inline notify.js]: Yes this is what I'm trying to explain, and why the regular events won't work.

Comment (perrin4869) [inline notify.js]: but just by reading the MDN documentation that is not clear, since `NotificationEvent.notification` is documented as returning a regular `Notification` object which, in theory, should support the events...

Comment (alexgibson) [inline notify.js]: It says `EventNotification.notification` is a read only property of the NotificationEvent interface, which is not what you're doing here? See the code example in the doc I linked to.

Comment (alexgibson) [inline notify.js]: I think it might be a good idea to rework this so the notifications are handled only in the SW, as we discussed. Perhaps then things might start to make a bit more sense in regard to how events are handled.

**Comment by perrin4869** [comment]
Well, the idea behind this pull request is to abstract the different APIs (Notification API and `serviceWorkerRegistration.showNotification`) as much as possible, so that you can do:

```javascript
const myNotification = new Notify("Title", { ... });
myNotification.show();
```

And not have to worry about the underlying API being used. Did you have something else in mind?

**Comment by alexgibson** [comment]
Well, it seems the approach your taking means that Notifications need to be handled in the web page itself, and not the service worker script? Won't this mean that if the application is closed, notifications won't really work as they should? See https://github.com/alexgibson/notify.js/issues/56#issuecomment-205004404

**Comment by perrin4869** [comment]
hm... need to do some more reading on service workers for sure

**Comment by perrin4869** [comment]
let me close this for now and send a new pull request from a separate branch
